### ISSUE TITLE
Passing meta data to winston custom logger

### ISSUE THREAD

**Issue by zoell** [balderdashy/captains-log#16]
Using a custom logger like:
```javascript
var customLogger = new winston.Logger({
    transports: [
        new(winston.transports.DailyRotateFile)({
            level: 'debug',
	    datePattern: '.yyyy-MM-dd',
	    json: false,
            filename: './logs/web.log',
    	    formatter: function(options) {
    	    return options.timestamp + ' ' + options.level.toUpperCase() + ' ' + options.meta.customentry + ' ' + (undefined !== options.message ? options.message : '');
    	    }
        }),
    ],
});
```
and setting this as noted in the readme:
```javascript
var log = require('captains-log')({ custom: customLogger });
log('hello', {customentry: 'metadata test'});
```
It seems this is not passing the meta to winston, it just concatenates to the 'hello' as a string.

Any idea?

**Comment by gotmikhail** [comment]
Same issue.

**Comment by gotmikhail** [comment]
https://github.com/balderdashy/sails/issues/2695#issuecomment-93071206

**Comment by Globegitter** [comment]
@zoell @gotmikhail if any of you two has a fix at hand I would be happy to merge it in since I think @mikermcneil has too much at hand right now to actively maintain this repo.

**Comment by japel** [comment]
captains log seems to pass the string concatenated at 
https://github.com/balderdashy/captains-log/blob/master/lib/write.js#L78

```
    // Compose `str` of all the arguments
    // (include the appropriate prefix if specified)
    var prefixStr = (options.prefixes && options.prefixes[logAt]) || '';
    var str = prefixStr + util.format.apply(util, pieces);

		// Call log fn
		return logFn.apply(logFn, [str]);
```
and winston trys to get metadata out of arguments passed to it at
https://github.com/winstonjs/winston/blob/master/lib/winston/logger.js#L128

```
  var callback = typeof args[args.length - 1] === 'function' ? args.pop() : null,
      meta     = typeof args[args.length - 1] === 'object' && Object.prototype.toString.call(args[args.length - 1]) !== '[object RegExp]' ? args.pop() : {},
      msg      = util.format.apply(null, args);
```
checking if its typeof object
captains log should pass the arguments to to winston and not a concatenated string

**Comment by sailsbot** [comment]
Thanks for posting, @zoell.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by Casen** [comment]
Why is this closed? It is still a nightmare situation to use a custom logger in sails. I absolutely hate that captains-log is a requirement in the first place. Is there anything on the docket for this? It seems like an easy fix.

**Comment by Casen** [comment]
Currently, captains-log must be used, but there is still an issue where sails.log will instantiate multiple loggers when you use a custom logger.

Also, captains-log concats everything into a string, which royally f*cks with log parsing. 

@japel outlined the fix back in September, so if no one else wants to implement it I would be happy to.

**Comment by dinana** [comment]
This should not be closed, this is still very much a pain in the neck.

**Comment by kevinob11** [comment]
I'd love to see this fixed as well, winston is misbehaving badly when I try to connect it with sails.

**Comment by mikermcneil** [comment]
Hey everyone- I added the custom log as a feature for folks who might need it, but we haven't been using that in production ourselves, so I haven't run up against this personally.  I'd be happy to merge the fix as outlined by @japel.  

Re: sailsbot- now that the `bug` label is here, she shouldn't try to close it again.  Sorry about that!  I've [heard](https://github.com/balderdashy/sails/issues/3423#issuecomment-169751072) that Sailsbot had to go away to finishing school to learn some more advanced etiquette, but that she will be back.

**Comment by mikermcneil** [comment]
> Is there anything on the docket for this? It seems like an easy fix.

@kevinob11 Here's what I think would make the most sense:

```
// Whenever you use:
sails.log.silly(x,y,z,...);
sails.log.verbose(x,y,z,...);
sails.log.info(x,y,z,...);
sails.log.debug(x,y,z,...);
sails.log.warn(x,y,z,...);
sails.log.error(x,y,z,...);
sails.log(x,y,z,...);


// The configurable custom log function receives:
function theCustomLogFn (logLevel, logArgs){
  // do custom stuff, e.g.
  if (logLevel === 'debug') {
    console.log.apply(console, logArgs);
  }
  else {
    // etc.
  }
}
```

To me, that seems like the best way to provide the flexible amount of flexibility and have sails get out of the way.  Would that work for everyone?

**Comment by mikermcneil** [comment]
> Currently, captains-log must be used, but there is still an issue where sails.log will instantiate multiple loggers when you use a custom logger.

@Casen To clarify, captains-log is used for the calls to `sails.log` in core.  Obviously you can avoid using `sails.log` in your own code and e.g. `require('winston')` that way.  But the best thing would be for a more flexible custom log function.  Would the interface I posted above do the trick?

**Comment by kevinob11** [comment]
After posting this I found that it was a bit easier to override than I thought and seems to work great. I used the code below and captain's log seems to be getting out of the way as long as I set `inspect: false`.

```
var winston = require('winston');
var customLogger = new winston.Logger();

customLogger.add(winston.transports.Console, {
  level: 'debug',
  colorize: true
});

customLogger.add(winston.transports.File, {
  level: 'error',
  filename: 'filename.log',
  json: false
});

module.exports.log = {

  custom: customLogger,
  level: 'silly',
  prefixThemes: {
    praece: {
      silly: '',
      verbose: '',
      info: '',
      blank: '',
      debug: '',
      warn: '',
      error: '',
      crit: ''
    }
  },
  prefixTheme: 'praece',
  inspect: false

};
```

**Comment by mikermcneil** [comment]
@kevinob11 :+1 glad to hear it (and thanks for sharing your solution!)

@Casen @japel @gotmikhail @zoell @dinana is this a workable solution for you guys?  The changes I suggested above would be doable in a backwards compatible way (much like we've done with other config in Sails-- just use a different prop name and respect the old if present+auto-adjust+log a deprecation message; if both are present, error on lift until fixed; and if the new is present all good).

However in the short term, I think this is a documentation issue, first and foremost.  Currently in the Sails [reference docs](http://sailsjs.org/documentation/reference/configuration/sails-config-log), the only official config exposed in Sails is the log `level` (even though technically we support the other configuration options mentioned here).  If any of you folks has a spare moment to document the other settings [on the relevant page of the sails-docs repo](https://github.com/balderdashy/sails-docs/blob/master/reference/sails.config/sails.config.log.md), that'd be great.  @rachaelshaw or I will merge with the other doc improvements going into v0.12.  Thanks!

@kevinob11 If you don't mind, I think it'd be great to include your code here in a note below the main config reference table as an example of how to do a complete override.

**Comment by kevinob11** [comment]
Sure I could give docs a shot, and I'm happy to have my example on there. I'll work on it this weekend.

**Comment by mikermcneil** [comment]
Thanks @kevinob11!

**Comment by japel** [comment]
@Casen I'd be happy if you did, as I have no idea what I meant or did back then :)
@mikermcneil sounds good to me... That doc issue is something you see a lot in sails or other sails related packages as they relay on or are built on other packages and just pass sails configs around... (e.g. sails-mysql, sails-mongo -> sails.config.connections) The philosophical question is to what extent document something or not (assuming people will go and read the documentation of the dependendy)

**Comment by mikermcneil** [comment]
@japel totally-- that's a perfect example.  And I feel like we keep running into that in general in Waterline-- but also in Skipper.  The dilemma is we either (A) document everything twice and have it context-specific with a better experience for users using the package directly (but way more maintenance work and likelihood of inconsistencies- as @uncletammy correctly pointed out back in 2014) or (B) we just lump it together in Sails docs (at the cost of slightly reduced utility or increased awkwardness for direct users of the lower level modules) or (C) lump it together in docs for the relevant module (e.g. Waterline or captains-log)-- which has all the side-effects as (B) but flip-flopped.

Increasingly, I'm leaning towards (B), because it helps out the most people and keeps the maintenance burden to a minimum.  In the case of captains-log, it's not like it's really all that properly documented in this repo either, so we're not really losing anything.  And I think as a general approach, linking to the appropriate section of Sails docs with a contextual message, e.g.:

> The configuration for captains-log is passed in as a dictionary of options as shown in the example above.  To see all available options, view the [reference page on log configuration on sailsjs.org](http://sailsjs.org/documentation/reference/configuration/sails-config-log#?properties))

**Comment by mikermcneil** [comment]
@japel and actually, the same approach would work well for public methods I think (except stuff that's specifically for direct usage outside of Sails-- e.g. in the case of Waterline, stuff like `Collection.extend()`)

**Comment by japel** [comment]
@mikermcneil agree on B being the best option of those... :+1:

**Comment by kevinob11** [comment]
Ok, I added the parameters needed to use a custom logger along with a working example of using winston. balderdashy/sails-docs#615 take a look and let me know.

**Comment by mikermcneil** [comment]
Thanks again @kevinob11-- just merged.  This will appear on the Sails website as soon as we double-check changes, merge master in 0.11 and redeploy (@rachaelshaw was sick today, but it's on her radar for this week).

@zoell I'm going to go ahead and close this issue, but if anything else comes up, please open a new issue.  Thanks!
